src/main/java/com/fasterxml/jackson/databind/ser/std/ReferenceTypeSerializer.java

```diff
@@ -408,17 +410,14 @@ public void serializeWithType(T ref,
         //    (which is what non-polymorphic serialization does too), we will need
         //    to simply delegate call, I think, and NOT try to use it here.
 
-        // Otherwise apply type-prefix/suffix, then std serialize:
-        /*
-        typeSer.writeTypePrefixForScalar(ref, g);
-        serialize(ref, g, provider);
-        typeSer.writeTypeSuffixForScalar(ref, g);
-        */
+        // [modules-java8#86] Since 2.19.0, Bringing back the prefix and suffix writing part
+        WritableTypeId typeId = typeSer.writeTypePrefix(g, typeSer.typeId(ref, JsonToken.VALUE_STRING));
```

Comment:
Ok. I think this makes sense, except for one problem -- `JsonToken.VALUE_STRING` works for some cases, not for others. Problem is, we do not actually know the expected shape of contained value. Later on, test can be broken; I'll add a note where and how, below.

Reply:
... I am really not sure how to fix this. I understand why type id would be needed, for `AtomicReference` / `Optional` wrapper. Unfortunately there is a key problem: nothing else will be written for these wrappers.
Unless... unless I guess if we decide that for the specific case of polymorphic value, there is.

And I think the only possibility, there, is extra wrapper array. I'll see if that works.

Reply:
Sigh. Nope. Was not able to make this work overall.

Reply:
At least if we have more test case 👍🏼👍🏼

Reply:
@cowtowncoder I modified the test via https://github.com/FasterXML/jackson-databind/pull/4838/commits/de925686e54db60c91e4010934c71f0a796d0a51, to configure `StdTypeResolverBuilder` to use **appropriate inclusion scheme** for each mapper. Seems like that's way to go?

Reply:
Perhaps? I am not sure which cases to focus most; ones with `@JsonTypeInfo` introduced ones or Default Typing (I know, it's reported wrt Default Typing but in a way `@JsonTypeInfo` cases are more important -- Default Typing having security concerns).

Another confounding factor is that for `String` and `Integer`, type id MUST NOT be written ("natural" types), nor expected.
It's a mess. :)

Reply:
But no matter what, we really cannot use `JsonToken.VALUE_STRING` for all cases because some values are serialized as Objects (Maps, most POJOs), others as Arrays (`Collection`s) and those cases would not work with this type id.

Yet there is also no way for `ReferecenTypeSerializer` to know for sure value shape.

This is why I _think_ we actually must use JSON Array wrapper for Type Id case to know exact shape for just this case (in theory JSON Object but that would require bogus key for single entry; array seems cleaner).

Reply:
I am starting to think this situation (Default Typing overrides`@JsonTypeInfo`) is inevitable, due to Jackson does heavy lifiting during serialization.
I don't think there is easy, sensible way to introspect all the way from `Foo<T>` -> `AtomicRference<T>` -> `Animal` to get configured type information, during runtime.

So maybe we can either...

1. separate concerns, resolve very specific issue here around `Optional<String>` and create separate milestone for this override case by pushing some tests as `tofix`.. or
2. Just add all tests to `tofix` and come back later in the future when we are in best shape to do so.

Reply:
I think the problem is more with the "missing" serialization for `Optional`, actually. For non-polymorphic case it is sort of ok, as long as POJO typing is defined. Basically following 2 serialize to identical JSON:

```
static class Required {
   public MyPojo value = new MyPojo();
}

static class Optional {
  public Optional<MyPojo> value = Optional.of(new MyPojo());
}

// Both become something like:

{ "value": {
   "myPojoProp1": 1,
   "myPojoProp2" : 2
}

// Compared to say

static class Multiple {
   public List<MyPojo> values = List.of(new MyPojo()_'
}

// which becomes

{ "values": [
   {
   "myPojoProp1": 1,
   "myPojoProp2" : 2
  }
]}
```

this is ok because no Type Id needs to be added anywhere.
But if (polymorphic) Type Id needs to be added, there is only place for one of Type Ids (`MyPojo`) -- for `Optional` there is no JSON value to modify/include because it is invisible, unlike `List` and others for which there is JSON Array.
If there was JSON Array to represent `Optional` level things would be easier.

Now: in theory we could consider doing something similar to `EnumSet` where Class name as type id is actually polymorphic -- and we might be able to do the same with optional. But it would have to replace Type Id of thing contained (and only works for Class Name type id, but that's probably to be expected).
But I suspect that would be even more work.

As to "solving" reported case... I am still worried I don't fully understand the behavior before and after change. We'd need to, I think, verify exact JSON being produce and not simply checking round-tripping.

Reply:
True, this is getting bigger than I imagined I agree with your concerns wrt serialization output. Knowing that the serialization output changed versions back, (2.8 -> 2.9) complicates things I think.

We might even consider current output as expected behavior and suggest to try work around or something, for cases where they update [ 2.8 -> 2.9 or later like 2.18 .

In any case, I will close this off for now and would like to get back to taking steps toward 3.0 -- I will leave some closing comments